### PR TITLE
Better version diff printing

### PR DESCRIPTION
**Pull request by Tyilo**
Before the common version string of `2.11.4` and `2.12.3` would be `2.1`.

With this patch it is now `2.`.

I don't know if we would prefer to get `2.` or just `2` as the common version string.
I'm also not sure if I'm missing any version delimiters.

### CODE REVIEW

**Comment by Tyilo** [comment]
I also simplified the code for the initial comparison to 3 lines: https://github.com/Tyilo/pikaur/blob/fb4f484280c8677ae2a2e1d069b87c219b6a2fbd/pikaur/pprint.py#L74-L76

File: pikaur/pprint.py [hunk 1]
```diff
@@ -68,20 +68,16 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
...
-                counter < len(str1)
-        ) and (
-            counter < len(str2)
-        ) and (
-            str1[counter] == str2[counter]
-        ):
-            result += str1[counter]
-            counter += 1
+    def get_common_version(str1, str2):
+        version_delimiters = ':.-+'
```
Comment (actionless) [inline pikaur/pprint.py]: typo

File: pikaur/pprint.py [hunk 2]
```diff
@@ -68,20 +68,16 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
...
+
+        result = str2
+        while not str1.startswith(result):
+            result = result[:-1]
+
+        while len(result) > 0 and result[-1] not in version_delimeters:
```
Comment (actionless) [inline pikaur/pprint.py]: typo

**Comment by actionless** [comment]
also this looks not consistent:

```py
In [2]: get_common_version('0.50.2-1', '0.50.2+6+123131-1')
Out[2]: '0.50.'

In [3]: get_common_version('0.50.2+1', '0.50.2+6+123131-1')
Out[3]: '0.50.2+'
```

File: pikaur/pprint.py [hunk 3]
```diff
@@ -68,20 +68,16 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
...
-        ) and (
-            str1[counter] == str2[counter]
-        ):
-            result += str1[counter]
-            counter += 1
+    def get_common_version(str1, str2):
+        version_delimiters = ':.-+'
+
+        result = str2
+        while not str1.startswith(result):
```
Comment (actionless) [inline pikaur/pprint.py]: what about `    while not str1.startswith(result.rstrip(version_delimeters)):`?

File: pikaur/pprint.py [hunk 4]
```diff
@@ -68,20 +68,16 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
...
+            result = result[:-1]
+
+        while len(result) > 0 and result[-1] not in version_delimeters:
+            result = result[:-1]
+
         return result
```
Comment (actionless) [inline pikaur/pprint.py]: `return result.rstrip(version_delimeters)` ?

**Comment by actionless** [comment]
hm, i've tried with few more values and seems what rstrip not helping in some cases

you can use this as a testcase:
```py
get_common_version('0.50.1', '0.50.1')
get_common_version('0.50.12', '0.50.13')
get_common_version('0.50.2-1', '0.50.2+6+123131-1')
get_common_version('0.50.2+1', '0.50.2+6+123131-1')
```

i think the best approach will be to split versions by all possible separators, ie:
`"0"."50"."2"-"1" -> ["0", "50" , "2" , "1"]`
`"0.50.2+6+123131-1" -> ["0", "50", "2", "6", "123131", "1"]`

and next operate with those lists instead of iterating through string characters

**Comment by Tyilo** [comment]
I'll try to come up with something better :)

**Comment by actionless** [comment]
i wasn't reviewing that really but i already see what print formatting happends in version.py intead of pprint, that's not fine because we can't know the context, so versions are printed the same on the left and the right:

2018-02-22--1519328455_1208x579_scrot

2018-02-22--1519328472_1294x551_scrot

File: pikaur/pprint.py [hunk 5]
```diff
@@ -115,16 +94,8 @@ def pretty_format(pkg_update):
             pkg_len += len(pkg_update.Repository) + 1
         return ' {pkg_name}{spacing} {current_version}{spacing2} -> {new_version}{verbose}'.format(
             pkg_name=pkg_name,
-            current_version=_color_line(common_version, version_color) +
```
Comment (actionless) [inline pikaur/pprint.py]: i think you should revert this change chunk and so remove the method `diff_print` from PackageVersion class

File: pikaur/version.py [hunk 6]
```diff
@@ -1,40 +1,171 @@
...
+        assert r1 == -r2
+        assert v1 <= v2
+
+        if r1 == 0:
+            print('%s == %s' % (v1.diff_print(), v2.diff_print()))
+        else:
+            print('%s < %s' % (v1.diff_print(p1), v2.diff_print(p2)))
+
+    # This should be sorted
+    versions = '''
```
Comment (actionless) [inline pikaur/version.py]: could you please refactor it to be a tuple of tuples?
```py
(
   (r23.3f69224-1, r24.fffffff-1),
(123, 125),
)
```
and so on

**Comment by actionless** [comment]
tonight i'll look into this more detailed

**Comment by actionless** [comment]
in order to address #27 i've refactored pikaur.version to use `vercmp` from pacman, please update your PR to use it

File: pikaur/version.py [hunk 7]
```diff
@@ -1,5 +1,125 @@
...
+                part = int(match[1])
+            else:
+                part = match[2]
+
+            span = match.span()
+            parts.append((part, (offset + span[0], offset + span[1])))
+
+        return parts
+
+    def compare(self, other):
```
Comment (actionless) [inline pikaur/version.py]: please use `pikaur.version.compare_versions` instead of reimplementing comparison again

File: .pylintrc [hunk 8]
```diff
@@ -50,7 +50,7 @@ confidence=
 # --enable=similarities". If you want to run only the classes checker, but have
 # no Warning level messages displayed, use"--disable=all --enable=classes
 # --disable=W"
-disable=too-few-public-methods,missing-docstring
+disable=too-few-public-methods,missing-docstring,too-many-return-statements
```
Comment (actionless) [inline .pylintrc]: if there is a real reason to break codestyle, add inline pylint hint iunstead of modifying pylint config

File: pikaur/version.py (hunk 7 above)
Comment (Tyilo) [reply pikaur/version.py]: We might as well implement it as it doesn't add any complexity to the code and it can be used in tests to check that it is consistent with `vercmp`.

I think the best solution is to rename the function to `_compare` and add the following method:

```
def common_prefix(self, other):
    _, prefix = self._compare(other)
    return self.version_string[:prefix]
```

and use that. We can then use `_compare` in the test code.

File: .pylintrc (hunk 8 above)
Comment (Tyilo) [reply .pylintrc]: Fixed

File: pikaur/version.py (hunk 7 above)
Comment (actionless) [reply pikaur/version.py]: i think it will be better to keep as simple as possible, so implement only a dedicated function for version diff, but not to do any extra stuff

Comment (Tyilo) [reply pikaur/version.py]: Literally the only difference to the code would be that `compare` would only return one value instead of a pair. We would still need exactly the same number of lines in `compare`.

How about doing what I said above, but also remove the `__str__` and `__lt__`, `__eq__`, etc. methods?

File: pikaur/pprint.py (hunk 5 above)
Comment (actionless) [reply pikaur/pprint.py]: and this one still not addressed

File: pikaur/version.py (hunk 7 above)
Comment (actionless) [reply pikaur/version.py]: so i would like to see that PR to be much smaller with just one new function for getting common part of the string, because less code always better in both terms of support and mainainability  -- i am more happy to merge PRs which have more red lines than green ones ;)

Comment (Tyilo) [reply pikaur/version.py]: The problem is that the common string between 2 versions isn't just a common prefix.

If the first version is `1.2.3-4` and the second one is `1.02.003-5` then the prefixes would be

```
[1.2.3]-4    =>    [1.02.003]-5
```

We can of course still do this with a single version, but then we would need to parse the two version strings twice so in total 4 parsings instead of the 2 we do currently.

File: pikaur/pprint.py (hunk 5 above)
Comment (Tyilo) [reply pikaur/pprint.py]: How is this not addressed?

File: pikaur/version.py (hunk 7 above)
Comment (actionless) [reply pikaur/version.py]: i think we need just to compare splitted versions by equality of blocks

Comment (Tyilo) [reply pikaur/version.py]: You forgot `:`, but we might as well do it properly like `pacman`/`vercmp` does it.

Comment (actionless) [reply pikaur/version.py]: in that case instead of just comparing them for equality we can pass each block to `compare_versions` func, ie to vercmp

File: pikaur/pprint.py (hunk 5 above)
Comment (actionless) [reply pikaur/pprint.py]: my bad, didn't pulled the local copy, now this is fine

File: pikaur/version.py (hunk 7 above)
Comment (Tyilo) [reply pikaur/version.py]: That wouldn't actually work as `1.2` and `1.2-5` should compare equal, whereas `1.2` and `1.2.5` shouldn't.

Also `1:2.3` and `1.2.3` shouldn't compare equal either.

File: pikaur/pprint.py [hunk 9]
```diff
@@ -68,41 +69,19 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
...
+        _, prefix1 = current_version.compare(new_version)
+        _, prefix2 = new_version.compare(current_version)
+
         version_color = 10
         old_color = 11
         new_color = 9
         column_width = min(int(get_term_width() / 2.5), 37)
         sort_by = '{:03d}{}'.format(
-            len(common_version)*10+len(pkg_update.New_Version),
+            prefix1*10+len(pkg_update.New_Version),
```
Comment (actionless) [inline pikaur/pprint.py]: you should sort not by length of common string but by number of blocks

i think it will be better to completely rewrite it all again and and just replace `get_common_string` and `get_version_diff` to one function `get_common_version(version1, version2)` which will return common version part (as string) and length of common blocks

File: pikaur/version.py (hunk 7 above)
Comment (actionless) [reply pikaur/version.py]: i think it should be enough just to additionally compare separator itself for equality

Comment (Tyilo) [reply pikaur/version.py]: I don't understand why we can't just do it correctly.

Do I really have to fork `pikaur` if I want correct version diffing behavior?

Comment (actionless) [reply pikaur/version.py]: tried my suggestion:

2018-02-25--1519531649_1327x1888_scrot

Comment (actionless) [reply pikaur/version.py]: whole patch is that small:

```diff
diff --git a/pikaur/pprint.py b/pikaur/pprint.py
index 2168f95..c6958c3 100644
--- a/pikaur/pprint.py
+++ b/pikaur/pprint.py
@@ -2,6 +2,7 @@ import sys
 import shutil
 
 from .config import VERSION
+from .version import compare_versions
 
 
 PADDING = 4
@@ -61,6 +62,9 @@ def print_not_found_packages(not_found_packages):
         print(format_paragraph(package))
 
 
+VERSION_SEPARATORS = ('.', '+', '-', ':')
+
+
 def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False, color=True):
 
     _color_line = color_line
@@ -68,21 +72,32 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
     if not color:
         _color_line = _bold_line = lambda x, *args: x
 
-    def get_common_string(str1, str2):
-        result = ''
-        if '' in (str1, str2):
-            return result
-        counter = 0
-        while (
-                counter < len(str1)
-        ) and (
-            counter < len(str2)
-        ) and (
-            str1[counter] == str2[counter]
-        ):
-            result += str1[counter]
-            counter += 1
-        return result
+    def split_version(version):
+        splitted_version = []
+        block = ''
+        for char in version:
+            if char in VERSION_SEPARATORS:
+                splitted_version.append(block)
+                splitted_version.append(char)
+                block = ''
+            else:
+                block += char
+        return splitted_version
+
+    def get_common_version(version1, version2):
+        common_string = ''
+        common_length = 0
+        if '' in (version1, version2):
+            return common_string, common_length
+        splitted_version1 = split_version(version1)
+        splitted_version2 = split_version(version2)
+        for block1, block2 in zip(splitted_version1, splitted_version2):
+            if compare_versions(block1, block2) == 0:
+                common_string += block1
+                common_length += 1
+            else:
+                break
+        return common_string, common_length
 
     def get_version_diff(version, common_version):
         new_version_postfix = version
@@ -94,7 +109,7 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
         return new_version_postfix
 
     def pretty_format(pkg_update):
-        common_version = get_common_string(
+        common_version, difference_size = get_common_version(
             pkg_update.Current_Version, pkg_update.New_Version
         )
         version_color = 10
@@ -102,7 +117,7 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
         new_color = 9
         column_width = min(int(get_term_width() / 2.5), 37)
         sort_by = '{:03d}{}'.format(
-            len(common_version)*10+len(pkg_update.New_Version),
+            difference_size,
             pkg_update.Name
         )
         pkg_name = _bold_line(pkg_update.Name)
```

Comment (actionless) [reply pikaur/version.py]: is any of the diffs is looking incorrect on the example screenshot?

Comment (Tyilo) [reply pikaur/version.py]: It looks fine, however I would still prefer my solution.

**Comment by Tyilo** [comment]
Closing as @actionless preferred solution has been added.
